### PR TITLE
tests: add a smoke test

### PR DESCRIPTION
**Pull request by bboozzoo**
Add a smoke spread test.

### CODE REVIEW

File: spread.yaml
Comment (bboozzoo) [inline spread.yaml]: 2024

File: spread.yaml
Comment (bboozzoo) [inline spread.yaml]: drop?

File: spread.yaml
Comment (bboozzoo) [inline spread.yaml]: drop?

File: tests/lib/nested.sh
Comment (bboozzoo) [inline tests/lib/nested.sh]: drop, unused

File: tests/lib/nested.sh
Comment (bboozzoo) [inline tests/lib/nested.sh]: drop, unused

File: tests/lib/tools/vmcli/setup.cfg
Comment (bboozzoo) [inline tests/lib/tools/vmcli/setup.cfg:13]: this has python packaging only so that we'll get auto install of telnetlib3 when installing the tool with pip

File: tests/lib/tools/vmcli/vmcli/main.py
Comment (zyga) [inline tests/lib/tools/vmcli/vmcli/main.py]: You can replace `Any` with `argparse.Namespace`, I believe. It has some smarts to understand that better AFAIR.

File: tests/lib/tools/imcmp.py
Comment (zyga) [inline tests/lib/tools/imcmp.py]: Trivial `argparse.Namespace`

File: tests/lib/tools/imcmp.py
Comment (zyga) [inline tests/lib/tools/imcmp.py]: `-> None`

File: tests/lib/prepare-utils.sh
Comment (zyga) [inline tests/lib/prepare-utils.sh:55]: I think you can use `--reflink=auto` to maybe get a speed-up if the file system is compatible and the snap is large.

File: tests/lib/prepare-utils.sh
Comment (zyga) [inline tests/lib/prepare-utils.sh:27]: `DEBIAN_FRONTENT=noninteractive`?

File: tests/lib/prepare-uc.sh
Comment (zyga) [inline tests/lib/prepare-uc.sh:114]: Nitpick, you can use `s,foo,bar,` to avoid some of the escape, perhaps?

Comment (zyga) [reply tests/lib/prepare-uc.sh:114]: I would love if we had vsock support for looking into test VMs that don't have network and have locked down ssh/serial/what-not.

File: tests/lib/prepare-uc.sh
Comment (zyga) [inline tests/lib/prepare-uc.sh:88]: Just use `tar -xf ... -C /`

File: tests/lib/nested.sh
Comment (zyga) [inline tests/lib/nested.sh:117]: I may have something for this to share.

**Review by zyga — COMMENTED** [review]
Quick pass through several places. Thanks a lot for picking this enormous work. Don't leave me in cross distro all by myself if field consumes you ;-)

File: .github/workflows/snap.yaml
Comment (alfonsosanchezbeato) [inline .github/workflows/snap.yaml:73]: Do we need a retention policy? I'm curious on whether this is storage we are really spending or if github does not care.

File: spread.yaml
Comment (alfonsosanchezbeato) [inline spread.yaml]: `image: ubuntu-24.04-64` seems to be working, at least for core-initrd tests

File: tests/lib/nested.sh
Comment (alfonsosanchezbeato) [inline tests/lib/nested.sh:1]: At some point we should really unify this between core-initrd&snaps&...

File: tests/lib/tools/imcmp.py
Comment (alfonsosanchezbeato) [inline tests/lib/tools/imcmp.py:1]: This should have a general description of what is it for

File: tests/main/smoke/task.yaml
Comment (alfonsosanchezbeato) [inline tests/main/smoke/task.yaml:62]: You probably don't want your user in the merged PR :slightly_smiling_face: - maybe we could use `snapd-test-brand-store@canonical.com` or other user we already have, o create a new one.

File: tests/lib/tools/vmcli/vmcli/main.py
Comment (alfonsosanchezbeato) [inline tests/lib/tools/vmcli/vmcli/main.py:1]: This would need also a description on what it does in the header.

**Review by alfonsosanchezbeato — COMMENTED** [review]
This looks very cool!

I have some comments, but also I am concerned about duplicating things that we have elsewhere. Concretely:
* Would it b possible to use `action-test` from `system-snaps-cicd-tools` instead of creating a new action?
* The scripts that run a nested VM would ideally reuse what we have in core-initrd, maybe we should add those scripts to a common place. Eventually even sharing code with snapd, but in that case there is a lot of history that makes reusing the scripts harder.

File: .github/workflows/snap.yaml
Comment (bboozzoo) [reply .github/workflows/snap.yaml:73]: IIRC the default is 90 day or so, which counts towards the overall organization quota. Snaps aren't big, but then again those artifacts aren't useful once the branch lands

File: tests/main/smoke/task.yaml
Comment (bboozzoo) [reply tests/main/smoke/task.yaml:62]: we need an email account associated with a LP user who uploaded their ssh keys

**Comment by bboozzoo** [comment]
> This looks very cool!
> 
> I have some comments, but also I am concerned about duplicating things that we have elsewhere. Concretely:
> 
>     * Would it b possible to use `action-test` from `system-snaps-cicd-tools` instead of creating a new action?

action-test is used already https://github.com/snapcore/console-conf-snap/pull/18/files#diff-95557d53bf91069e59d82b9d8fcfaf52ec84a762858983edd5ef3c9b3e4c8191L22 the PR enables spread test with it. I've added a 2nd independent worfklow which uses snapcore/action-build and spread tests because it completes significantly faster than LP builds :smile: (especially with armhf rebuild of the archive wait times were hitting the 6h timeout)

On a side note, I wonder if it's maybe be useful to split action-test to a piece which builds on LP (checkout, build, upload-artifact steps) and a another piece which runs spread tests (download-artifact, run spread tests, discard steps). This way it'd be easier to rerun the failed steps if the spread workflow wasn't super stable.
> 
>     * The scripts that run a nested VM would ideally reuse what we have in core-initrd, maybe we should add those scripts to a common place. Eventually even sharing code with snapd, but in that case there is a lot of history that makes reusing the scripts harder.

Yes, let's do that. I'll file a backlog ticket for it.

File: .github/workflows/snap.yaml
Comment (alfonsosanchezbeato) [reply .github/workflows/snap.yaml:73]: Got it, thanks

File: tests/main/smoke/task.yaml
Comment (alfonsosanchezbeato) [reply tests/main/smoke/task.yaml:62]: We could upload some, and put the private key in a github secret so the system can be accessed via ssh even and make the test more complete. But maybe for the future. If we do we should also make the private key available in a readable place for anybody in the team.

File: tests/main/smoke/task.yaml
Comment (alfonsosanchezbeato) [inline tests/main/smoke/task.yaml:91]: Maybe make it a loop to reduce the line count, and print the concrete image that was different.

File: tests/main/smoke/task.yaml
Comment (alfonsosanchezbeato) [inline tests/main/smoke/task.yaml:54]: I see that some screenshots are taken but not compared, which I think is fine, but maybe it is worth adding a TODO to check also those screens in the future.

Also, I think these are copied to artifacts, is that correct?

**Review by alfonsosanchezbeato — APPROVED** [review]
Thanks for the changes. I have some minor comments in case you want to address, but otherwise LGTM.

File: tests/main/smoke/task.yaml
Comment (bboozzoo) [reply tests/main/smoke/task.yaml:54]: yes, artifacts are always collected whether the task succeeds or fails, the artifact-upload step runs always and  will make them available for inspection even if the test failed

File: tests/main/smoke/task.yaml
Comment (bboozzoo) [reply tests/main/smoke/task.yaml:91]: I'll land this one and open a followup PR, just to exercise the CI loop a bit more :smile:

**Pull request merged**
